src/TDB2.cpp

```diff
@@ -42,967 +42,161 @@
-{
-  // Ensure the task is consistent, and provide defaults if necessary.
-  // bool argument to validate() is "applyDefault". Pass add_to_backlog through
-  // in order to not apply defaults to synchronized tasks.
-  task.validate (add_to_backlog);
-  std::string uuid = task.get ("uuid");
 
-  // If the tasks are loaded, then verify that this uuid is not already in
-  // the file.
-  if (!verifyUniqueUUID (uuid))
-    throw format ("Cannot add task because the uuid '{1}' is not unique.", uuid);
+  auto ws = replica.working_set ();
 
-  // Only locally-added tasks trigger hooks.  This means that tasks introduced
-  // via 'sync' do not trigger hooks.
-  if (add_to_backlog)
-    Context::getContext ().hooks.onAdd (task);
+  // get the ID that was assigned to this task
+  auto id = ws.by_uuid (uuid);
 
-  update (task, add_to_backlog, true);
-}
-
-////////////////////////////////////////////////////////////////////////////////
-void TDB2::modify (Task& task, bool add_to_backlog /* = true */)
-{
-  // Ensure the task is consistent.
-  task.validate (false);
-  std::string uuid = task.get ("uuid");
+  // update the cached working set with the new information
+  _working_set = std::make_optional (std::move (ws));
 
-  // Get the unmodified task as reference, so the hook can compare.
-  if (add_to_backlog)
-  {
-    Task original;
-    get (uuid, original);
-    Context::getContext ().hooks.onModify (original, task);
+  if (id.has_value ()) {
+    task.id = id.value();
   }
 
-  update (task, add_to_backlog);
+  // run hooks for this new task
+  Context::getContext ().hooks.onAdd (task);
 }
 
 ////////////////////////////////////////////////////////////////////////////////
-void TDB2::update (
-  Task& task,
-  const bool add_to_backlog,
-  const bool addition /* = false */)
+// Modify the task in storage to match the given task.
+//
+// This exhibits a bit of a race condition: if the stored task has changed since
+// it was loaded, this will revert those changes.  In practice, this is not an
+// issue.
+void TDB2::modify (Task& task, bool/* = true */)
 {
-  // Validate to add metadata.
   task.validate (false);
+  auto uuid = task.get ("uuid");
 
-  // If the task already exists, it is a modification, else addition.
+	// invoke the hook and allow it to modify the task before updating
   Task original;
-  if (not addition && get (task.get ("uuid"), original))
-  {
-    // Update only if the tasks differ
-    if (task == original)
-      return;
-
-    if (add_to_backlog)
-    {
-      // All locally modified tasks are timestamped, implicitly overwriting any
-      // changes the user or hooks tried to apply to the "modified" attribute.
-      task.setAsNow ("modified");
+  get (uuid, original);
+	Context::getContext ().hooks.onModify (original, task);
+
+  auto maybe_tctask = replica.get_task (uuid);
+  if (!maybe_tctask.has_value ()) {
+    throw std::string ("task no longer exists");
+  }
```

Comment:
It's mostly defensive programming. The DB uses sequential transactions, but the Taskwarrior code itself is not executed within that transaction. So it's possible for a task to be modified by some other process (such as if the user is running multiple `task` processes at the same time) between being read and the call to TDB2::modify. In practice, even if that occurs the task is unlikely to no longer exist, since we don't delete tasks until they expire.